net/usbwlan.c

```diff
@@ -0,0 +1,521 @@
+		.bRequest = cmd,
+		.wValue = (uint16_t)(regaddr & 0xffff),
+		.wIndex = (uint16_t)((regaddr >> 16) & 0xffff),
+		.wLength = size,
+	};
+	return usb_transferControl(dev->drv, dev->pipeCtrl, &setup, data, size, usb_dir_in);
+}
+
+
+static int usbwlan_bulkSend(usbwlan_dev_t *dev, const void *data, size_t size)
+{
+	return usb_transferBulk(dev->drv, dev->pipeBulkOUT, (void *)data, size, usb_dir_out);
+}
+
+
+static int usbwlan_bulkReceive(usbwlan_dev_t *dev, void *data, size_t size)
+{
+	return usb_transferBulk(dev->drv, dev->pipeBulkIN, data, size, usb_dir_in);
+}
+
+
+static int usbwlan_handleDevCtl(msg_t *msg, usbwlan_dev_t *dev)
+{
+	usbwlan_i_t *imsg = (usbwlan_i_t *)msg->i.raw;
+
+	switch (imsg->type) {
+		case usbwlan_dl:
+			return usbwlan_dlCmd(dev, imsg->dl.cmd, imsg->dl.wIndex, msg->o.data, msg->o.size);
+
+		case usbwlan_ctrl_in:
+			return usbwlan_ctrlReceive(dev, msg->o.data, msg->o.size);
+
+		case usbwlan_ctrl_out:
+			return usbwlan_ctrlSend(dev, msg->i.data, msg->i.size);
+
+		case usbwlan_reg_read:
+			return usbwlan_regRead(dev, imsg->reg.cmd, imsg->reg.regaddr, msg->o.data, msg->o.size);
+
+		case usbwlan_reg_write:
+			return usbwlan_regWrite(dev, imsg->reg.cmd, msg->i.data, msg->i.size);
+
+		default:
+			return -EINVAL;
+	}
+}
+
+
+static void usbwlan_msgthread(void *arg)
+{
+	usbwlan_dev_t *dev;
+	msg_t msg;
+	msg_rid_t rid;
+
+	for (;;) {
+		if (msgRecv(usbwlan_common.msgport, &msg, &rid) < 0) {
+			fprintf(stderr, "usbwlan: msgRecv returned with err\n");
+			break;
+		}
+
+		/* Ignore this msg, as it might have been sent by us after deletion event */
+		if (msg.type == mtUnlink) {
+			msg.o.err = EOK;
+			msgRespond(usbwlan_common.msgport, &msg, rid);
+			continue;
+		}
+
+		dev = usbwlan_get(msg.oid.id);
+		if (dev == NULL) {
+			msg.o.err = -ENOENT;
+			msgRespond(usbwlan_common.msgport, &msg, rid);
+			continue;
+		}
+
+		/* A device can be opened only by one process */
+		/* FIXME: allow mtClose with different PID (files closing on process exit have invalid PID value as of 2023-10-23) */
+		if (((msg.type != mtOpen) && (msg.type != mtClose)) && (msg.pid != dev->clientpid)) {
+			usbwlan_put(dev);
+			msg.o.err = -EBUSY;
+			msgRespond(usbwlan_common.msgport, &msg, rid);
+			continue;
+		}
+
+		switch (msg.type) {
+			case mtOpen:
+				if (dev->clientpid != 0) {
+					msg.o.err = -EBUSY;
+					break;
+				}
+				msg.o.err = usbwlan_open(dev, msg.i.openclose.flags, msg.pid);
+				break;
+
+			case mtClose:
+				usbwlan_close(dev);
+				msg.o.err = EOK;
+				break;
+
+			case mtRead:
+				msg.o.err = usbwlan_bulkReceive(dev, msg.o.data, msg.o.size);
+				break;
+
+			case mtWrite:
+				msg.o.err = usbwlan_bulkSend(dev, msg.i.data, msg.i.size);
+				break;
+
+			case mtDevCtl:
+				msg.o.err = usbwlan_handleDevCtl(&msg, dev);
+				break;
+
+			case mtDestroy:
+				usbwlan_put(dev);
+				msgRespond(usbwlan_common.msgport, &msg, rid);
+				endthread();
+				return;
```

Comment:
`endthread()` should not return. Maybe replace `return` with `__builtin_unreachable()` to signify that instead

(we should mark `endthread` as `noreturn` at last)